xonsh/ptk_shell/shell.py

```diff
@@ -240,50 +262,99 @@ def cmdloop(self, intro=None):
                 else:
                     break
 
-    def prompt_tokens(self):
-        """Returns a list of (token, str) tuples for the current prompt."""
-        p = builtins.__xonsh__.env.get("PROMPT")
+    def update_prompt(self, prompt_name: str, toks_futures: Tuple[List[str], dict]):
+        """Listen on futures and update the prompt as each one is completed.
+
+        :param prompt_name : what prompt to update. One of ['message', 'rprompt', 'bottom_toolbar']
+        :param toks_futures : a tuple containing tokens and futures for that prompt.
+        """
+
+        tokens, futures = toks_futures
+        timer = None
+
+        for fut in concurrent.futures.as_completed(futures):
+            val = fut.result() or ""
+
+            if fut not in futures:
+                # rare case where the future is completed but the container is already cleared
+                # because new prompt is called
+                continue
+
+            token_index = futures[fut]
+            if isinstance(token_index, int):
+                tokens[token_index] = val
+            else:  # when the function is called outside shell.
+                for idx, sect in enumerate(tokens):
+                    if token_index in sect:
+                        tokens[idx] = sect.replace(token_index, val)
+
+            # calling invalidate in less period is inefficient
+            timer = self._invalidate_prompt(timer, tokens, prompt_name)
+
+    def _invalidate_prompt(self, timer, tokens: List[str], prompt_name: str):
+        """Create a timer to update the prompt. The timing can be configured through env variables.
+            threading.Timer is used to stop calling invalidate frequently.
+        """
+        if timer:
+            timer.cancel()
+
+        def _invalidate():
+            new_prompt = "".join(tokens)
+            formatted_tokens = tokenize_ansi(
+                PygmentsTokens(partial_color_tokenize(new_prompt))
+            )
+            setattr(self.prompter, prompt_name, formatted_tokens)
+            self.prompter.app.invalidate()
+
+        timer = threading.Timer(
+            builtins.__xonsh__.env["ASYNC_INVALIDATE_INTERVAL"], _invalidate
+        )
+        timer.start()
+        return timer
+
+    def _get_prompt_tokens(self, env_name: str, prompt_name: str, **kwargs):
+        env = builtins.__xonsh__.env
+        p = env.get(env_name)
+
+        if not p and "default" in kwargs:
+            return kwargs.pop("default")
+
         try:
-            p = self.prompt_formatter(p)
+            template = p
+            if env["ENABLE_ASYNC_PROMPT"]:
+                self.prompt_updators[prompt_name] = self.prompt_formatter(
+                    template=template, threaded=True
+                )
+                p = "".join(self.prompt_updators[prompt_name][0])
+            else:
+                p = self.prompt_formatter(template=template)
```

Comment:
please try to limit the try-block to a single line.  Feel free to make a helper function if needed

Reply:
done.